chore: granular public simulation benchmarks

just-mitch opened the pull request:
Allow the PXE to send classes that get registered to its node to allow better debug logging

Consolidate the kernel phases enum
Add benchmarking stats for AVM simulation
Add benchmarking stats for public DB accesses made by avm simulation

alexghr commented on yarn-project/foundation/src/types/index.ts:
```diff
@@ -15,3 +15,12 @@ export type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
 export type Writeable<T> = { -readonly [P in keyof T]: T[P] };
+
+// Only works for up to 4 parts separated by hyphens
+export type ToCamelCase<S extends string> = S extends `${infer P1}-${infer P2}-${infer P3}-${infer P4}`
+  ? `${P1}${Capitalize<P2>}${Capitalize<P3>}${Capitalize<P4>}`
+  : S extends `${infer P1}-${infer P2}-${infer P3}`
+  ? `${P1}${Capitalize<P2>}${Capitalize<P3>}`
+  : S extends `${infer P1}-${infer P2}`
+  ? `${P1}${Capitalize<P2>}`
+  : S;
```
😆

alexghr commented on yarn-project/circuit-types/src/stats/metrics.ts:
```diff
@@ -122,6 +122,48 @@ export const Metrics = [
     description: 'Time for the sequencer to execute public function calls for txs in a block.',
     events: ['l2-block-built'],
   },
+  {
+    name: 'setup_kernel_simulation_avg_ms',
+    groupBy: 'block-size',
```
What do you think about having a generic `public_kernel_simulation_ms` metric groupings by `kernel_type` (i.e. `setup`, `app`, `teardown`, `tail`). I don't see what benefit we're getting from separating these stats by block size, maybe I'm missing something.

just-mitch replied:
Nope, you're missing nothing. I agree.

alexghr commented on yarn-project/simulator/src/public/benchmarker.ts:
```diff
@@ -0,0 +1,54 @@
+    return result;
+  }
+
+  getAverageDurations(): AverageDurations {
```
The stats module already averages all the values for a metric so I think this leads to an average of averages (which most likely isn't what we want?).

https://github.com/AztecProtocol/aztec-packages/blob/39747b933a13aa08f25c5074207f9d92489d5e3d/yarn-project/scripts/src/benchmarks/aggregate.ts#L295-L305

alexghr commented on yarn-project/simulator/src/public/benchmarker.ts:
```diff
@@ -0,0 +1,54 @@
+
+  async benchmark<T>(phase: PublicKernelPhase, event: Event, callback: () => T) {
+    const [durationMs, result] = await elapsed(callback);
+    this.durations[`${phase}-${event}`].push(durationMs);
```
Highly recommend the new histogram in Node

https://nodejs.org/dist/latest-v18.x/docs/api/perf_hooks.html#perf_hookscreatehistogramoptions

alexghr commented on yarn-project/circuit-types/src/stats/metrics.ts:
```diff
@@ -122,6 +122,48 @@ export const Metrics = [
+  },
+  {
+    name: 'setup_avm_simulation_avg_ms',
+    groupBy: 'block-size',
```
I think the AVM simulation stats need to be grouped by function name as they depend on the size/complexity of the bytecode being executed.

just-mitch committed:
Allow the PXE to send classes that get registered to its node to allow better debug logging
Consolidate the kernel phases enum
Add benchmarking stats for AVM simulation
Add benchmarking stats for public DB accesses made by avm simulation
